src/libcore/ptr.rs

```diff
@@ -821,6 +1281,708 @@ impl<T: ?Sized> *mut T {
+    ///
+    /// # Safety
+    ///
+    /// This operation is marked unsafe because it accepts a raw pointer.
+    ///
+    /// It does not drop the contents of `self`. This is safe, but it could leak
+    /// allocations or resources, so care must be taken not to overwrite an object
+    /// that should be dropped.
+    ///
+    /// This is appropriate for initializing uninitialized memory, or overwriting
+    /// memory that has previously been `read` from.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// #![feature(pointer_methods)]
+    ///
+    /// let mut x = 0;
+    /// let y = &mut x as *mut i32;
+    /// let z = 12;
+    ///
+    /// unsafe {
+    ///     y.write_volatile(z);
+    ///     assert_eq!(y.read_volatile(), 12);
+    /// }
+    /// ```
+    #[unstable(feature = "pointer_methods", issue = "43941")]
+    #[inline]
+    pub unsafe fn write_volatile(self, val: T)
+        where T: Sized,
+    {
+        write_volatile(self, val)
+    }
+
+    /// Overwrites a memory location with the given value without reading or
+    /// dropping the old value.
+    ///
+    /// Unlike `write`, the pointer may be unaligned.
+    ///
+    /// # Safety
+    ///
+    /// This operation is marked unsafe because it writes through a raw pointer.
+    ///
+    /// It does not drop the contents of `self`. This is safe, but it could leak
+    /// allocations or resources, so care must be taken not to overwrite an object
+    /// that should be dropped.
+    ///
+    /// Additionally, it does not drop `src`. Semantically, `src` is moved into the
+    /// location pointed to by `dst`.
+    ///
+    /// This is appropriate for initializing uninitialized memory, or overwriting
+    /// memory that has previously been `read` from.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// #![feature(pointer_methods)]
+    ///
+    /// let mut x = 0;
+    /// let y = &mut x as *mut i32;
+    /// let z = 12;
+    ///
+    /// unsafe {
+    ///     y.write_unaligned(z);
+    ///     assert_eq!(y.read_unaligned(), 12);
+    /// }
+    /// ```
+    #[unstable(feature = "pointer_methods", issue = "43941")]
+    #[inline]
+    pub unsafe fn write_unaligned(self, val: T)
+        where T: Sized,
+    {
+        write_unaligned(self, val)
+    }
+
+    /// Replaces the value at `self` with `src`, returning the old
+    /// value, without dropping either.
+    ///
+    /// # Safety
+    ///
+    /// This is only unsafe because it accepts a raw pointer.
+    /// Otherwise, this operation is identical to `mem::replace`.
+    #[unstable(feature = "pointer_methods", issue = "43941")]
+    #[inline]
+    pub unsafe fn replace(self, src: T) -> T
+        where T: Sized,
+    {
+        replace(self, src)
+    }
+
+    /// Swaps the values at two mutable locations of the same type, without
+    /// deinitializing either. They may overlap, unlike `mem::swap` which is
```

Comment:
By 'they may overlap' you mean 'the locations may be equal'?

What happens if you swap the values at, say, two different but overlapping locations (at least one of the pointers will be unaligned)? Is it allowed to call `swap` with unaligned pointers?

Reply:
You don't need alignment issues -- consider `x: [u32; 3]`, where I swap [u32; 2] from x[0] and x[1].

Reply:
In this case the final value of x[1] is actually unspecified, but definitely one of the original values of x[0] or x[2].